apiserver/test/e2e/service_server_e2e_test.go

```diff
@@ -225,6 +225,131 @@ func TestGetAllServices(t *testing.T) {
 	require.Equal(t, tCtx.GetNamespaceName(), response.Services[0].Namespace)
 }
 
+func TestGetAllServicesWithPagination(t *testing.T) {
+	const numberOfNamespaces = 3
+	const numberOfService = 2
+	const totalServices = numberOfNamespaces * numberOfService
+
+	tCtxs := make([]*End2EndTestingContext, numberOfNamespaces)
+	expectedServiceNames := make([][]string, numberOfNamespaces)
+
+	// Create services for each namespace
+	for i := 0; i < numberOfNamespaces; i++ {
+		tCtx, err := NewEnd2EndTestingContext(t)
+		require.NoError(t, err, "No error expected when creating testing context")
+
+		tCtx.CreateComputeTemplate(t)
+		t.Cleanup(func() {
+			tCtx.DeleteComputeTemplate(t)
+		})
+
+		expectedServiceNames[i] = make([]string, 0, numberOfService)
+		for j := 0; j < numberOfService; j++ {
+			testServiceRequest := createTestServiceV2(t, tCtx)
+			t.Cleanup(func() {
+				tCtx.DeleteRayService(t, testServiceRequest.Service.Name)
+			})
+			expectedServiceNames[i] = append(expectedServiceNames[i], testServiceRequest.Service.Name)
+		}
+
+		tCtxs[i] = tCtx
+	}
+
+	var pageToken string
+	tCtx := tCtxs[0]
+	// Test pagination with limit 1, which is less than the total number of services in all namespaces.
+	t.Run("Test pagination return part of the result services", func(t *testing.T) {
+		pageToken = ""
+		gotServices := make([][]bool, numberOfNamespaces)
+		for i := 0; i < numberOfNamespaces; i++ {
```

Comment:
suggest to add some comments for the 2d array, not easy to figure out at first glance